[HUDI-2277] HoodieDeltaStreamer reading ORC files directly using ORCDFSSource

zhangyue19921010 opened the pull request:
https://issues.apache.org/jira/projects/HUDI/issues/HUDI-2277

## What is the purpose of the pull request
Develop a new Source named ORCDFSSource extended from RowSource

Now, HoodieDeltaStreamer can read orc files directly using ORCDFSSource.

Also add UTs which are necessary and tested on our local env.

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

zhangyue19921010 commented:
@hudi-bot run travis

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/util/AvroOrcUtils.java:
```diff
@@ -796,4 +800,32 @@ private static Schema getActualSchemaType(Schema unionSchema) {
     }
   }
+
+  public static void addAvroRecord(
```
can this sit somewhere in a test utils class? given its only used by a test?

zhangyue19921010 replied:
Sure, Done.

vinothchandar commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/functional/TestHoodieDeltaStreamer.java:
```diff
@@ -1435,6 +1476,26 @@ public void testParquetDFSSourceWithSchemaFilesAndTransformer() throws Exception
+  }
+
+  @Test
+  public void testORCDFSSourceWithSourceSchemaFileAndNoTransformer() throws Exception {
```
for sake of test runtime. could we test just 1-2 combos here. most of the testing done for row source (ParquetDFS) should cover already?

zhangyue19921010 replied:
Okay, done. Thanks a lot for your review :)

vinothchandar reviewed (approved):
Just some small comments. We can land once they are addressed

zhangyue19921010 commented:
@hudi-bot run azure

zhangyue19921010 commented:
Hi @vinothchandar Thanks a lot for your review. Now Travis and Azure are all passed. PTAL :)

zhangyue19921010 commented:
Hi @vinothchandar sorry to bother you. Since this patch is passed all ut/it. So could you please take a look at your convince？ Thanks a lot!

zhangyue19921010 commented:
Or could we land it if possible? :)

vinothchandar commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/functional/TestHoodieDeltaStreamer.java:
```diff
@@ -1398,6 +1399,46 @@ private void testParquetDFSSource(boolean useSchemaProvider, List<String> transf
+            PROPS_FILENAME_TEST_ORC, ORC_SOURCE_ROOT, false);
+  }
+
+  private void prepareORCDFSSource(boolean useSchemaProvider, boolean hasTransformer, String sourceSchemaFile, String targetSchemaFile,
```
wonder if we can simplify this test?

zhangyue19921010 replied:
Srue thing. done. thanks a lot for your review.

zhangyue19921010 commented:
@hudi-bot run azure

nsivabalan commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/functional/TestHoodieDeltaStreamer.java:
```diff
@@ -1398,6 +1399,34 @@ private void testParquetDFSSource(boolean useSchemaProvider, List<String> transf
+                    transformerClassNames, PROPS_FILENAME_TEST_ORC, false,
+                    useSchemaProvider, 100000, false, null, null, "timestamp", null), jsc);
+    deltaStreamer.sync();
+    TestHelpers.assertRecordCount(ORC_NUM_RECORDS, tableBasePath + "/*/*.parquet", sqlContext);
```
shouldn't this be *.orc instead of *.parquet

zhangyue19921010 replied:
Hi @nsivabalan Thanks for your review. I think this is .parquet Because this patch is a ORCDFSSource which let HoodieDeltaStreamer can read orc file into hudi table and also use parquet format as base file format. So that we need to use .parquet when reading hudi table data.

nsivabalan replied:
he he. my bad, got it. thanks.

nsivabalan commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/testutils/UtilitiesTestBase.java:
```diff
@@ -364,5 +385,32 @@ public static String toJsonString(HoodieRecord hr) {
       return records.stream().map(Helpers::toJsonString).toArray(String[]::new);
     }
+
+    public static void addAvroRecord(
```
does this need to be public? can we switch to private.

zhangyue19921010 replied:
Sure thing. changed.

nsivabalan commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/testutils/UtilitiesTestBase.java:
```diff
@@ -364,5 +385,32 @@ public static String toJsonString(HoodieRecord hr) {
+
+      batch.size++;
+
+      if (batch.size % orcBatchSize == 0 || batch.size == batch.getMaxSize()) {
```
can you help me understand what this code block is doing? I see this method is called for one batch of records. lets say there are 100 records in one batch. if I am not wrong, batch.size at the end of adding 100 records should be 100. and so batch.size % orcBatchSize will be equal to 0 only after adding all records. 
If thats the case, shouldn't we move this block outside of this method. 
or am I missing something.

zhangyue19921010 replied:
Sure thing, changed :)

nsivabalan commented:
@zhangyue19921010 : do ping me here once you have addressed all comments. I can take a look.

zhangyue19921010 commented:
> @zhangyue19921010 : do ping me here once you have addressed all comments. I can take a look.

Hi @nsivabalan Thanks a lot for your review. I finished all the changes. PTAL :)

zhangyue19921010 commented:
Hi @nsivabalan @vinothchandar. Thanks a lot for your attention, review and approve! Could we land it or what else do I need to do?  :)

xushiyan commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/functional/TestHoodieDeltaStreamerBase.java:
```diff
@@ -247,4 +254,27 @@ protected static void prepareParquetDFSFiles(int numRecords, String baseParquetP
+    if (useCustomSchema) {
+      Helpers.saveORCToDFS(Helpers.toGenericRecords(
+              dataGenerator.generateInsertsAsPerSchema("000", numRecords, schemaStr),
+              schema), new Path(path), AvroOrcUtils.createOrcSchema(HoodieTestDataGenerator.AVRO_TRIP_SCHEMA));
```
better if add a `HoodieTestDataGenerator.ORC_TRIP_SCHEMA` in the class and convert this inside?

xushiyan commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/testutils/UtilitiesTestBase.java:
```diff
@@ -314,6 +320,28 @@ public static void saveParquetToDFS(List<GenericRecord> records, Path targetFile
     }
 
+    public static void saveORCToDFS(List<GenericRecord> records, Path targetFile) throws IOException {
+      TypeDescription orcSchema = AvroOrcUtils.createOrcSchema(HoodieTestDataGenerator.AVRO_SCHEMA);
```
ditto

xushiyan commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/functional/TestHoodieDeltaStreamer.java:
```diff
@@ -1622,6 +1651,16 @@ public void testParquetDFSSourceWithSchemaFilesAndTransformer() throws Exception
     testParquetDFSSource(true, Collections.singletonList(TripsWithDistanceTransformer.class.getName()));
   }
 
+  @Test
+  public void testORCDFSSourceWithoutSchemaProviderAndNoTransformer() throws Exception {
+    testORCDFSSource(false, null);
+  }
+
+  @Test
+  public void testORCDFSSourceWithSchemaFilesAndTransformer() throws Exception {
+    testORCDFSSource(true, Collections.singletonList(TripsWithDistanceTransformer.class.getName()));
+  }
```
can we use `@ParameterizedTest` here? with `@MethodSource` returning `Stream<Arguments>` to make it cleaner

xushiyan reviewed (approved):
LGTM. some minor code suggestions. Please go ahead and merge once fixed and Azure CI passed. Thanks.

xushiyan commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieTestDataGenerator.java:
```diff
@@ -129,10 +129,12 @@
       HoodieAvroUtils.addMetadataFields(AVRO_SCHEMA);
   public static final Schema AVRO_SHORT_TRIP_SCHEMA = new Schema.Parser().parse(SHORT_TRIP_SCHEMA);
   public static final Schema AVRO_TRIP_SCHEMA = new Schema.Parser().parse(TRIP_SCHEMA);
+  public static final Schema ORC_TRIP_SCHEMA = new Schema.Parser().parse(TRIP_SCHEMA);
```
actually what i meant is to do `AvroOrcUtils.createOrcSchema()` here for this constant so callers can use it directly

xushiyan replied:
@zhangyue19921010 I think I wasn't clear about the suggestion. What i meant is, here in HoodieTestDataGenerator, 

```suggestion
  public static final TypeDescription ORC_TRIP_SCHEMA = AvroOrcUtils.createOrcSchema(new Schema.Parser().parse(TRIP_SCHEMA));
```

This constant is named as `ORC_XXX_SCHEMA` but its type still an avro schema, causing confusion. That's why i suggest do conversion here, and make the constant less confusing and easier to use.

xushiyan commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/functional/TestHoodieDeltaStreamer.java:
```diff
@@ -1622,6 +1652,12 @@ public void testParquetDFSSourceWithSchemaFilesAndTransformer() throws Exception
     testParquetDFSSource(true, Collections.singletonList(TripsWithDistanceTransformer.class.getName()));
   }
 
+  @ParameterizedTest
+  @MethodSource("testArguments")
```
these 2 should be annotated on `testORCDFSSource()` method itself then this method not needed

xushiyan commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/functional/TestHoodieDeltaStreamer.java:
```diff
@@ -1936,4 +1972,12 @@ public Schema getTargetSchema() {
     }
   }
 
+  private static Stream<Arguments> testArguments() {
```
`testArguments` is not specific enough. use the same method name then `@MethodSource` does not need to have its argument. it'll look for the method with the same name. so this argument provider method will only serve its own test case method

```suggestion
  private static Stream<Arguments> testORCDFSSource() {
```

xushiyan commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieTestDataGenerator.java:
```diff
@@ -129,10 +129,12 @@
 
 
   public static final Schema AVRO_SCHEMA = new Schema.Parser().parse(TRIP_EXAMPLE_SCHEMA);
+  public static final Schema ORC_SCHEMA = new Schema.Parser().parse(TRIP_EXAMPLE_SCHEMA);
```
ditto

```suggestion
  public static final TypeDescription ORC_SCHEMA = AvroOrcUtils.createOrcSchema(new Schema.Parser().parse(TRIP_EXAMPLE_SCHEMA));
```

xushiyan reviewed (commented):
@zhangyue19921010 we're almost there :) just 1 more nitpick. thanks for making the feature.

zhangyue19921010 commented:
Hi @xushiyan Thanks a lot for your attention and review. My bad for misunderstanding :) code changed and waiting for ci/cd green.
